File: app/src/main/java/sfllhkhan95/doodle/core/models/PaintCanvas.java
package sfllhkhan95.doodle.core.models;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.PointF;
import android.util.DisplayMetrics;

import sfllhkhan95.doodle.R;
import sfllhkhan95.doodle.projects.utils.DoodleDatabase;


/**
 * @author saifkhichi96
 */
public class PaintCanvas extends Canvas {

    private static final int DEFAULT_BG_COLOR = R.color.defaultCanvasColor;
    private int bgColor;

    private Bitmap bgImage;
    private Bitmap bitmap;
    private String projectName = null;

    public PaintCanvas(Context context, DisplayMetrics metrics) {
        bgColor = context.getResources().getColor(DEFAULT_BG_COLOR);

        int height = metrics.heightPixels;
        int width = metrics.widthPixels;

        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565);
        setBitmap(bitmap);
    }

    public static PaintCanvas loadFromBitmap(Context context, DisplayMetrics metrics, Bitmap srcBmp) {
        PaintCanvas canvas = new PaintCanvas(context, metrics);

        float deviceAspect = metrics.widthPixels / (float) metrics.heightPixels;
        float bmpAspect = srcBmp.getWidth() / (float) srcBmp.getHeight();

        if (deviceAspect != bmpAspect) {
            if (deviceAspect > bmpAspect) { // Device is wider, fit width
                float targetHeight = srcBmp.getWidth() / deviceAspect;
                srcBmp = Bitmap.createBitmap(
                        srcBmp,
                        0,
                        (int) ((srcBmp.getHeight() - targetHeight) / 2),
                        srcBmp.getWidth(),
                        (int) targetHeight
                );
            } else {
                float targetWidth = srcBmp.getHeight() * deviceAspect;
                srcBmp = Bitmap.createBitmap(
                        srcBmp,
                        (int) ((srcBmp.getWidth() - targetWidth) / 2),
                        0,
                        (int) targetWidth,
                        srcBmp.getHeight()
                );
            }

            if (metrics.widthPixels > metrics.heightPixels) {
                int w = metrics.widthPixels;
                int h = (int) (metrics.widthPixels * srcBmp.getHeight() / (float) srcBmp.getWidth());
                canvas.bgImage = Bitmap.createScaledBitmap(srcBmp, w, h, false);
            } else {
                int h = metrics.heightPixels;
                int w = (int) (metrics.heightPixels * srcBmp.getWidth() / (float) srcBmp.getHeight());
                canvas.bgImage = Bitmap.createScaledBitmap(srcBmp, w, h, false);
            }
        } else {
            canvas.bgImage = Bitmap.createScaledBitmap(srcBmp, metrics.widthPixels, metrics.heightPixels, false);
        }

        return canvas;
    }

    public static PaintCanvas loadFromPath(Context context, DisplayMetrics metrics, String bmpPath) {
        Bitmap srcBmp = DoodleDatabase.loadDoodle(bmpPath, metrics.widthPixels, metrics.heightPixels);
        PaintCanvas canvas = loadFromBitmap(context, metrics, srcBmp);
        canvas.setProjectName(bmpPath);

        return canvas;
    }

    public int getColor() {
        return bgColor;
    }

    public void setColor(int bgColor) {
        this.bgColor = bgColor;
    }

    public Bitmap getBitmap() {
        return bitmap;
    }

    public void drawShapes(Tools tools) {
        tools.paint(this);
    }

    public void drawBackground() {
        if (bgImage == null) {
            drawColor(bgColor);
        } else {
            drawBitmap(bgImage, 0, 0, null);
        }
    }

    private void setProjectName(String title) {
        this.projectName = title;
    }

    public void clearProjectName() {
        this.projectName = null;
    }

    public void saveProject() {
        if (projectName == null) {
            DoodleDatabase.saveDoodle(bitmap);
        } else {
            DoodleDatabase.saveDoodle(bitmap, projectName);
        }
    }

    public int getColor(PointF touchAt) {
        try {
            return bitmap.getPixel((int) touchAt.x, (int) touchAt.y);
        } catch (IllegalArgumentException ex) {
            ex.printStackTrace();
            return DEFAULT_BG_COLOR;
        }
    }
}

File: app/src/main/java/sfllhkhan95/doodle/core/models/PaintCanvas.kt
package sfllhkhan95.doodle.core.models

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.PointF
import android.support.v4.content.ContextCompat
import android.util.DisplayMetrics
import sfllhkhan95.doodle.R
import sfllhkhan95.doodle.projects.utils.DoodleDatabase


/**
 * @author saifkhichi96
 */
class PaintCanvas(context: Context, metrics: DisplayMetrics) : Canvas() {
    var color: Int = 0

    private var bgImage: Bitmap? = null
    val bitmap: Bitmap
    private var projectName: String? = null

    init {
        color = ContextCompat.getColor(context, DEFAULT_BG_COLOR)

        val height = metrics.heightPixels
        val width = metrics.widthPixels

        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
        setBitmap(bitmap)
    }

    fun drawShapes(tools: Tools) {
        tools.paint(this)
    }

    fun drawBackground() {
        if (bgImage == null) {
            drawColor(color)
        } else {
            drawBitmap(bgImage!!, 0f, 0f, null)
        }
    }

    private fun setProjectName(title: String) {
        this.projectName = title
    }

    fun clearProjectName() {
        this.projectName = null
    }

    fun saveProject() {
        if (projectName == null) {
            DoodleDatabase.saveDoodle(bitmap)
        } else {
            DoodleDatabase.saveDoodle(bitmap, DoodleDatabase.timestamp2Name(projectName!!))
        }
    }

    fun getColor(touchAt: PointF): Int {
        return try {
            bitmap.getPixel(touchAt.x.toInt(), touchAt.y.toInt())
        } catch (ex: IllegalArgumentException) {
            ex.printStackTrace()
            DEFAULT_BG_COLOR
        }

    }

    companion object {

        private const val DEFAULT_BG_COLOR = R.color.blue_grey_500

        fun loadFromBitmap(context: Context, metrics: DisplayMetrics, source: Bitmap): PaintCanvas {
            var bitmap = source
            val canvas = PaintCanvas(context, metrics)

            val deviceAspect = metrics.widthPixels / metrics.heightPixels.toFloat()
            val bmpAspect = bitmap.width / bitmap.height.toFloat()

            if (deviceAspect != bmpAspect) {
                if (deviceAspect > bmpAspect) { // Device is wider, fit width
                    val targetHeight = bitmap.width / deviceAspect
                    bitmap = Bitmap.createBitmap(
                            bitmap,
                            0,
                            ((bitmap.height - targetHeight) / 2).toInt(),
                            bitmap.width,
                            targetHeight.toInt()
                    )
                } else {
                    val targetWidth = bitmap.height * deviceAspect
                    bitmap = Bitmap.createBitmap(
                            bitmap,
                            ((bitmap.width - targetWidth) / 2).toInt(),
                            0,
                            targetWidth.toInt(),
                            bitmap.height
                    )
                }

                if (metrics.widthPixels > metrics.heightPixels) {
                    val w = metrics.widthPixels
                    val h = (metrics.widthPixels * bitmap.height / bitmap.width.toFloat()).toInt()
                    canvas.bgImage = Bitmap.createScaledBitmap(bitmap, w, h, false)
                } else {
                    val h = metrics.heightPixels
                    val w = (metrics.heightPixels * bitmap.width / bitmap.height.toFloat()).toInt()
                    canvas.bgImage = Bitmap.createScaledBitmap(bitmap, w, h, false)
                }
            } else {
                canvas.bgImage = Bitmap.createScaledBitmap(bitmap, metrics.widthPixels, metrics.heightPixels, false)
            }

            return canvas
        }

        fun loadFromPath(context: Context, metrics: DisplayMetrics, bmpPath: String): PaintCanvas {
            val srcBmp = DoodleDatabase.loadDoodle(bmpPath, metrics.widthPixels, metrics.heightPixels)
            val canvas = loadFromBitmap(context, metrics, srcBmp!!)
            canvas.setProjectName(bmpPath)

            return canvas
        }
    }
}